flink-runtime/src/main/java/org/apache/flink/runtime/io/network/netty/NettyMessage.java

```diff
@@ -287,33 +317,37 @@ void releaseBuffer() {
 		@Override
 		ByteBuf write(ByteBufAllocator allocator) throws IOException {
 			checkNotNull(buffer, "No buffer instance to serialize.");
+			final int headerLength = 16 + 4 + 4 + 1 + 4;
 
-
-			ByteBuf result = null;
+			ByteBuf headerBuf = null;
 			try {
-				ByteBuffer nioBufferReadable = buffer.getNioBufferReadable();
-				int length = 16 + 4 + 4 + 1 + 4 + nioBufferReadable.remaining();
-
-				result = allocateBuffer(allocator, ID, length);
-
-				receiverId.writeTo(result);
-				result.writeInt(sequenceNumber);
-				result.writeInt(backlog);
-				result.writeBoolean(buffer.isBuffer());
-				result.writeInt(nioBufferReadable.remaining());
-				result.writeBytes(nioBufferReadable);
-
-				return result;
+				// in order to forward the buffer to netty, it needs an allocator set
+				buffer.setAllocator(allocator);
+
+				// only allocate header buffer - we will combine it with the data buffer below
+				headerBuf = allocateBuffer(allocator, ID, headerLength, buffer.readableBytes(), false);
+
+				receiverId.writeTo(headerBuf);
+				headerBuf.writeInt(sequenceNumber);
+				headerBuf.writeInt(backlog);
+				headerBuf.writeBoolean(buffer.isBuffer());
+				headerBuf.writeInt(buffer.readableBytes());
+
+				CompositeByteBuf composityBuf = allocator.compositeBuffer();
```

Comment:
nit: Maybe `compositeDirectBuffer(2)` or `compositeDirectBuffer(16) will be better, since it's more explicit? This default value of `16` for max number of components is quite a good candidate for a "gotcha of the month" some when in the future.

Reply:
we're going for the default now since we don't know how Netty is using it internally